mpi3mr/storcli2.go

```diff
@@ -0,0 +1,631 @@
+	Description     string `json:"Description"`
+}
+
+type DriveGroup struct {
+	DG      int    `json:"DG,omitempty"`
+	Array   int    `json:"Arr,omitempty"`
+	Row     int    `json:"Row,omitempty"`
+	EIDSlot string `json:"EID:Slot,omitempty"`
+	PID     int    `json:"PID,omitempty"`
+	DID     int    `json:"DID,omitempty"`
+	Type    string `json:"Type,omitempty"`
+	State   string `json:"State,omitempty"`
+	Status  string `json:"Status,omitempty"`
+	BT      string `json:"BT,omitempty"`
+	Size    string `json:"Size,omitempty"`
+	PDC     string `json:"PDC,omitempty"`
+	PI      string `json:"PI,omitempty"`
+	SED     string `json:"SED,omitempty"`
+	DS3     string `json:"DS3,omitempty"`
+	Secured string `json:"Secured,omitempty"`
+	FSpace  string `json:"FSpace,omitempty"`
+}
+
+func (dg *DriveGroup) UnmarshalJSON(data []byte) error {
+	tmp := make(map[string]interface{})
+	if err := json.Unmarshal(data, &tmp); err != nil {
+		return err
+	}
+	for k, v := range tmp {
+		switch k {
+		case "DG":
+			if v == "-" {
+				dg.DG = int(-1)
+			} else {
+				dg.DG = int(v.(float64))
+			}
+		case "Arr":
+			if v == "-" {
+				dg.Array = int(-1)
+			} else {
+				dg.Array = int(v.(float64))
+			}
+		case "Row":
+			if v == "-" {
+				dg.Row = int(-1)
+			} else {
+				dg.Row = int(v.(float64))
+			}
+		case "EID:Slot":
+			dg.EIDSlot = v.(string)
+		case "DID":
+			if v == "-" {
+				dg.DID = int(-1)
+			} else {
+				dg.DID = int(v.(float64))
+			}
+		case "Type":
+			dg.Type = v.(string)
+		case "State":
+			dg.State = v.(string)
+		case "BT":
+			dg.BT = v.(string)
+		case "Size":
+			dg.Size = v.(string)
+		case "PDC":
+			dg.PDC = v.(string)
+		case "PI":
+			dg.PI = v.(string)
+		case "SED":
+			dg.SED = v.(string)
+		case "DS3":
+			dg.DS3 = v.(string)
+		case "FSpace":
+			dg.FSpace = v.(string)
+		default:
+		}
+	}
+	return nil
+}
+
+type VirtualDrives struct {
+	VDInfo         VirtualDrive           `json:"VD Info"`
+	PhysicalDrives []PhysicalDrive        `json:"PDs"`
+	VDProperties   VirtualDriveProperties `json:"VD Properties"`
+}
+
+/*
+VD Info
+
+	"VD Info": {
+	  "DG/VD": "0/1",
+	  "TYPE": "RAID0",
+	  "State": "Optl",
+	  "Access": "RW",
+	  "CurrentCache": "NR,WB",
+	  "DefaultCache": "NR,WB",
+	  "Size": "893.137 GiB",
+	  "Name": ""
+	}
+*/
+type VirtualDrive struct {
+	DGVD           string                 `json:"DG/VD,omitempty"`
+	Type           string                 `json:"TYPE,omitempty"`
+	State          string                 `json:"State,omitempty"`
+	Access         string                 `json:"Access,omitempty"`
+	CurrentCache   string                 `json:"CurrentCache,omitempty"`
+	DefaultCache   string                 `json:"DefaultCache,omitempty"`
+	Size           string                 `json:"Size,omitempty"`
+	Name           string                 `json:"Name,omitempty"`
+	Properties     VirtualDriveProperties `json:"Properties,omitempty"`
+	PhysicalDrives []PhysicalDrive        `json:"PhysicalDrives,omitempty"`
+}
+
+/*
```

Comment:
looks like you ended up not needing newVirtualDrive, right? Looks like you get the VirtualDrive struct via unmarshalling input, and then assign the properties etc directly to that

Reply:
good catch, lemme drop.